src/com/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -928,119 +757,269 @@
+   ;; Ensure that all the required paths and values exist, and then
+   ;; insert the new facts.
+   (let [paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
+         pathstrs (map (comp facts/factpath-to-string first) paths-and-valuemaps)
+         valuemaps (map second paths-and-valuemaps)
+         vhashes (map :value_hash valuemaps)
+         path-ids (map (realize-paths! pathstrs) pathstrs)
+         vhash-ids (map :id (map (realize-values! valuemaps) vhashes))
+         pv-pairs (map vector path-ids vhash-ids)]
+     (insert-facts-pv-pairs! (certname-to-factset-id name) pv-pairs))))
+
+(defn-validated update-facts!
   "Given a certname, querys the DB for existing facts for that
    certname and will update, delete or insert the facts as necessary
-   to match the facts argument."
-  [{:keys [name values environment timestamp producer-timestamp] :as fact-data} :- facts-schema]
-  (let [factset (certname-to-factset-id name)
-        old-facts (current-fact-value-ids factset)
-        new-facts (new-fact-value-ids values)]
-    (sql/update-values :factsets ["id=?" factset]
-                       {:timestamp (to-timestamp timestamp)
-                        :environment_id (ensure-environment environment)
-                        :producer_timestamp (to-timestamp producer-timestamp)})
-    (utils/diff-fn (zipmap new-facts (repeat nil))
-                   (zipmap old-facts (repeat nil))
-                   #(insert-facts! factset %)
-                   #(delete-facts! factset %)
-                   identity)))
+   to match the facts argument. (cf. add-facts!)"
+  [{:keys [name values environment timestamp producer-timestamp] :as fact-data}]
+
+  ;; Might be possible to make some of this faster via sorted parallel traversal.
+  (sql/transaction
+   (let [factset-id (certname-to-factset-id name)
+         initial-factset-paths-vhashes
+         (sql/with-query-results result-set
+           ["SELECT fp.path, fv.value_hash FROM facts f
+               INNER JOIN fact_paths fp ON f.fact_path_id = fp.id
+               INNER JOIN fact_values fv ON f.fact_value_id = fv.id
+               WHERE factset_id = ?"
+            factset-id]
+           (doall result-set))
+         ;; Ensure that all the required paths and values exist.
+         paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
+         pathstrs (map (comp facts/factpath-to-string first) paths-and-valuemaps)
+         path-ids (map (realize-paths! pathstrs) pathstrs)
+         valuemaps (map second paths-and-valuemaps)
+         vhashes (map :value_hash valuemaps)
+         vhash-ids (map :id (map (realize-values! valuemaps) vhashes))]
+
+     ;; Add new facts and remove obsolete facts.
+     (let [replacement-pv-pairs (set (map vector path-ids vhash-ids))
+           current-pairs (set (select-pid-vid-pairs-for-factset factset-id))
+           new-pairs (set/difference replacement-pv-pairs current-pairs)
```

Comment:
@senior right - switched to data/diff instead of the dual set/difference.  I'd originally misunderstood the data/diff docs, and thought it was saying that the [path-id value-id] pairs would have to be sets for it to work correctly, not that the *parent* collection (of the pairs) just had to be a set.

After some trivial benchmarking I found that data/diff appeared be quite a bit faster from a cold start, and still notably faster on a second pass.  Thanks for pursuing this.